Send response for update request when it timed out

brwe opened the pull request:
otherwise the request will hang

nik9000 commented on core/src/test/java/org/elasticsearch/index/store/ExceptionRetryIT.java:
```diff
@@ -136,4 +144,58 @@ public void sendRequest(DiscoveryNode node, long requestId, String action, Trans
+                });
+            }
+        }
+        // crate index on one node and later send request through another
```
s/crate/create/

nik9000 commented on core/src/test/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -352,6 +353,17 @@ public String getClusterName() {
         return nodes.keySet().toArray(Strings.EMPTY_ARRAY);
     }
 
+    public String[] getDataNodeNames() {
```
Why not return the `List<String>`? Maybe some JavaDoc too.

nik9000 commented:
Left minor comments. This seems important to get into 2.0.0.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java:
```diff
@@ -202,7 +204,7 @@ public void handleException(TransportException exp) {
                         // we already marked it as started when we executed it (removed the listener) so pass false
                         // to re-add to the cluster listener
-                        retry(null);
+                        retry(exp.unwrapCause());
```
can we put this in a variable?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java:
```diff
@@ -214,6 +216,15 @@ public void handleException(TransportException exp) {
+                Throwable listenFailure = failure;
+                if (listenFailure == null) {
+                    if (shardIt == null) {
+                        listenFailure = new ElasticsearchTimeoutException("Timeout waiting for [" + request.timeout() + "], request: " + request.toString());
```
can we use the new fancy string formatting support in our Exceptions? Also, I'm not sure request.toString makes sense (I know it was how it was done), not all requests implement it  (I think!) and we will just have an ugly pointer. Maybe use the action name instead?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java:
```diff
@@ -214,6 +216,15 @@ public void handleException(TransportException exp) {
+                        listenFailure = new ElasticsearchTimeoutException("[" + shardIt.size() + "] shardIt, [" + shardIt.sizeActive() + "] active : Timeout waiting for [" + request.timeout() + "], request: " + request.toString());
+                    }
+                }
+                listener.onFailure(listenFailure);
```
also, can we remove the boolean return value from doStart and remove the timeout handling from the public void onTimeout(TimeValue timeout)  method of the callback given to the observer in line 245? just call doStart.

bleskes commented on core/src/test/java/org/elasticsearch/index/store/ExceptionRetryIT.java:
```diff
@@ -136,4 +144,58 @@ public void sendRequest(DiscoveryNode node, long requestId, String action, Trans
+     * Tests that an update request does not hang in case the request fails and times out.
+     */
+    @Test
+    public void testBulkUpdateWithObserverTimeout() throws ExecutionException, InterruptedException, IOException {
```
can we make a unit test out of this? it's TransportInstanceSingleOperationAction that's broken. Not the update request.

brwe replied:
yes, unit test would be better

bleskes commented:
Left comments...

brwe commented:
@nik9000 @bleskes thanks a lot for the review! I addressed all comments. I also found another bug: If `resolveRequest(..)` returned `false` then the request would also hang. Fixed and added a unit test for it.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java:
```diff
@@ -138,13 +141,14 @@ protected boolean doStart() {
                 // check if we need to execute, and if not, return
                 if (!resolveRequest(observer.observedState(), request, listener)) {
-                    return true;
+                    listener.onFailure(new IllegalStateException(LoggerMessageFormat.format("{} request {} could not be resolved", new ShardId(request.index, request.shardId), actionName)));
```
good catch

bleskes commented on core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java:
```diff
@@ -169,11 +173,11 @@ protected boolean doStart() {
+                return;
             }
 
             if (!operationStarted.compareAndSet(false, true)) {
```
do we still need this operationStarted?

brwe replied:
I am not sure. I thought it was an additional protection against duplicated messages? Say for example, handleException() in line 203 is called twice for whatever reason then we might end up updating twice?

bleskes replied:
I don't think we need to protected ourselves here against violation of transport service semantics (just one callback).

brwe replied:
ok, removed

bleskes commented on core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java:
```diff
@@ -203,18 +207,27 @@ public void handleException(TransportException exp) {
                         // we already marked it as started when we executed it (removed the listener) so pass false
                         // to re-add to the cluster listener
-                        retry(null);
+                        Throwable cause = exp.unwrapCause();
```
I meant assign it out of the if and use it in all places :)

bleskes commented on core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java:
```diff
@@ -232,17 +245,7 @@ public void onClusterServiceClose() {
-                        }
-                        listener.onFailure(listenFailure);
-                    }
+                    doStart();
```
++

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,249 @@
+        clusterService.setState(ClusterStateCreationUtils.state("test", randomBoolean(), ShardRoutingState.STARTED));
+        action.new AsyncSingleAction(request, listener).start();
+        assertThat(transport.capturedRequests().length, equalTo(1));
+        listener.onResponse(new Response());
```
don't we want to send this response back into the transport?

brwe replied:
indeed

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,249 @@
+        action.new AsyncSingleAction(request, listener).start();
+        assertThat(transport.capturedRequests().length, equalTo(1));
+        listener.onResponse(new Response());
+        listener.get(100, TimeUnit.MILLISECONDS);
```
why 100?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,249 @@
+        listener.get(100, TimeUnit.MILLISECONDS);
+    }
+
+    public void testRetryOnTimeout() throws InterruptedException, ExecutionException, TimeoutException {
```
can we randomize or separate tests for the cause of timeouts? we can have a connection / shard closed / primary unassigned etc?

brwe replied:
ok, I can add a few but not sure how much value that adds. I don't really test here if the retry was triggered or not

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,249 @@
+        assertThat(transport.capturedRequests().length, equalTo(0));
+        try {
+            listener.get(1000, TimeUnit.MILLISECONDS);
+        } catch (Throwable t) {
```
feels like we need to fail if we don't have an exception?

brwe replied:
yes

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,249 @@
+        action.new AsyncSingleAction(request, listener).start();
+        assertThat(transport.capturedRequests().length, equalTo(0));
+        try {
+            listener.get(1000, TimeUnit.MILLISECONDS);
```
why 1000? this should never hang right? we can just use get()? if it hangs it's an issue.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,249 @@
+        action.new AsyncSingleAction(request, listener).start();
+        assertThat(transport.capturedRequests().length, equalTo(0));
+        try {
+            listener.get(1000, TimeUnit.MILLISECONDS);
```
same comment as above about the use of a time?

brwe replied:
I made it so when writing the test so that it would not hang forever. But I can remove that now.

bleskes commented:
Thx @brwe - this looks great. Left some comments here and there..

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,280 @@
+            public <T extends TransportResponse> void sendRequest(DiscoveryNode node, String action, TransportRequest request, TransportRequestOptions options, TransportResponseHandler<T> handler) {
+                TransportException e;
+                int excpetionType = randomInt(4);
+                switch (excpetionType) {
```
partially replying to the original message - the idea is to test how this action behaves under different error conditions - a connection error (which is thrown here),  a general exception / shard not available (which is typically given as a response , not thrown here (although it's equivalent at the moment).

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,280 @@
+                        break;
+                    }
+                    default: {
+                        throw new RuntimeException("we should never get here!");
```
I think you mean this to fail the test. Does it?

brwe replied:
yes, with this stack trace: 	

```
    at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationActionTests$1.sendRequest(TransportInstanceSingleOperationActionTests.java:219)
	at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationAction$AsyncSingleAction.doStart(TransportInstanceSingleOperationAction.java:185)
	at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationAction$AsyncSingleAction.start(TransportInstanceSingleOperationAction.java:126)
	at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationActionTests.testRetryOnTimeout(TransportInstanceSingleOperationActionTests.java:238)
	
```

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,280 @@
+                        e = new TransportException("test exception", new NoShardAvailableActionException(new ShardId("test", 0)));
+                        break;
+                    }
+                    case 4: {
```
this test is about timeouts, right? feels unnatural to throw a normal exception here. You can also rename to the test to say "handle exceptions" and check all exception are handled correctly (timeout or thrown) - imho a deidcate non-timed out error is better (i.e., it tests it returns immediately and not after a short timeout).

brwe commented:
@bleskes I changed the tests now so that there is dedicated tests for immediate failure, locally triggered retry, triggered retry from remote etc. I like it better now but am sure you will find more :)

bleskes commented on core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java:
```diff
@@ -197,24 +196,32 @@ public void handleResponse(Response response) {
+                    if (cause instanceof ConnectTransportException || cause instanceof NodeClosedException ||
                             retryOnFailure(exp)) {
-                        operationStarted.set(false);
                         // we already marked it as started when we executed it (removed the listener) so pass false
```
this comment can go away now?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,317 @@
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        clusterService.setState(ClusterStateCreationUtils.state("test", randomBoolean(), ShardRoutingState.STARTED));
+
+        // start the action but then simulate a network disconnect
```
is the network disconnect comment correct now?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,317 @@
+
+        try {
+            // result should return immediately
+            listener.get();
```
do we want to check listener.isDone? (because it is supposed to return immediately). get() waits.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,317 @@
+        boolean local = randomBoolean();
+        clusterService.setState(ClusterStateCreationUtils.state("test", local, ShardRoutingState.STARTED));
+        action.new AsyncSingleAction(request, listener).start();
+        // this should fail because primary not initialized
```
copy paste issue?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,317 @@
+        request.shardId = 0;
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        clusterService.setState(ClusterStateCreationUtils.state("test", randomBoolean(), ShardRoutingState.STARTED));
+        // start the action but then simulate a network disconnect
```
same copy paste issue?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationActionTests.java:
```diff
@@ -0,0 +1,317 @@
+        requestId = transport.capturedRequests()[0].requestId;
+        transport.handleResponse(requestId, new ConnectTransportException(node, "test exception"));
+        try {
+            // result should return immediately
```
same comment on listener.isDone

bleskes commented:
This looks great. Thx @brwe for all the iterations. I left some minor comments. No need for another review imho.
